[Jetpack Content Migration Flow] Refactor - Unify user flags provider

mkevins opened the pull request:
# Description

This PR continues from #17398 by refactoring the user flags provider. With the unified entity provider, the entity-specific logic is extracted to a helper class for the provider-side.

### To test:

#### Precondition:

Turn on the various flags. The following patch may be useful for this purpose:

<details>
<summary>Patch to enable all data migration flags</summary>

```patch
diff --git a/WordPress/build.gradle b/WordPress/build.gradle
index 1a23203a63..cfa2e01bc0 100644
--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -120,11 +120,11 @@ android {
         buildConfigField "boolean", "BETA_SITE_DESIGNS", "false"
         buildConfigField "boolean", "JETPACK_POWERED", "true"
         buildConfigField "boolean", "JETPACK_POWERED_BOTTOM_SHEET", "false"
-        buildConfigField "boolean", "JETPACK_SHARED_LOGIN", "false"
-        buildConfigField "boolean", "JETPACK_LOCAL_USER_FLAGS", "false"
-        buildConfigField "boolean", "JETPACK_BLOGGING_REMINDERS_SYNC", "false"
-        buildConfigField "boolean", "JETPACK_READER_SAVED_POSTS", "false"
-        buildConfigField "boolean", "JETPACK_PROVIDER_SYNC", "false"
+        buildConfigField "boolean", "JETPACK_SHARED_LOGIN", "true"
+        buildConfigField "boolean", "JETPACK_LOCAL_USER_FLAGS", "true"
+        buildConfigField "boolean", "JETPACK_BLOGGING_REMINDERS_SYNC", "true"
+        buildConfigField "boolean", "JETPACK_READER_SAVED_POSTS", "true"
+        buildConfigField "boolean", "JETPACK_PROVIDER_SYNC", "true"
         buildConfigField "boolean", "JETPACK_MIGRATION_FLOW", "false"
         buildConfigField "boolean", "JETPACK_FEATURE_REMOVAL_PHASE_ONE", "false"
         buildConfigField "boolean", "JETPACK_FEATURE_REMOVAL_PHASE_TWO", "false"
```
</details>

Log into WordPress:

1. Install the WordPress app flavor from this branch
2. Login to the WordPress app and select a site
3. Make note of (or change some) user flags

Log out of Jetpack and delete all data:

* Force stop the app (if running)
* Go to storage and clear app data

#### Testing steps:

1. Run the Jetpack app from this branch
2. Expect to be logged in automatically
4. Expect to see any changed user flags copied to Jetpack



## Regression Notes
1. Potential unintended areas of impact
No. Existing implementations are changed, but are behind a feature flag.

9. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

10. What automated tests I added (or what prevented me from doing so)
These will be refactored in a later PR to keep the change set manageable.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mkevins committed:
Extract local content entity enum to separate file

mkevins committed:
Refactor user flags provider to unified provider

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/localcontentmigration/LocalContentEntity.kt:
```diff
@@ -0,0 +1,35 @@
+
+import org.wordpress.android.fluxc.model.PostModel
+
+enum class LocalContentEntity(private val isSiteContent: Boolean = false) {
```
I wanted to confirm if it makes sense to have a default value for `isSiteContent` as `false` here and in line 10 have a default value of `true`.

mkevins replied:
After our more recent discussions, with the shift in approach to preserve local ids (where possible), it may make sense to migrate all "site content" entities at once (and not site by site), so soon, we may no longer need to include a separate local site id within the URI. So, in a later PR, this field will likely be removed.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/userflags/resolver/UserFlagsResolver.kt:
```diff
@@ -34,35 +28,21 @@ class UserFlagsResolver @Inject constructor(
+                userFlagsAnalyticsTracker.trackFailed(ErrorType.NoUserFlagsFoundError)
+                onFailure()
+            }
+//            userFlagsAnalyticsTracker.trackFailed(ErrorType.QueryUserFlagsError)
```
Should we uncomment or remove this line before merging?

mkevins replied:
Handled w/ https://github.com/wordpress-mobile/WordPress-Android/pull/17423/commits/85553bb376299203fdab1ffd07a9aa40a945203c

RenanLukas reviewed (commented):
Hey @mkevins , thanks for the PR! I've left a couple of comments. Please let me know if they make sense.

mkevins committed:
Merge branch 'try/local-migration-content-provider--refactor-login-resolver-as-orchestrator' into try/local-migration-content-provider--refactor-unify-user-flags-provider

mkevins committed:
Merge branch 'trunk' into try/local-migration-content-provider--refactor-unify-user-flags-provider

mkevins committed:
Remove commented code

mkevins committed:
Merge branch 'trunk' into try/local-migration-content-provider--refactor-unify-user-flags-provider

mkevins committed:
Move user flag helper methods to helper and check for null

This moves the helper methods to the user flags helper class, and also
adds null checks to wrap the calls to WellSQL java code.

mkevins committed:
Update local content entity for sites list

mkevins committed:
Rename sites list entity to be plural

This also removes the local entity id parameter, since we can just pass
the entire list of site models. This should be ok for sites, since the
list should usually be small, and the model data is also small.

mkevins committed:
Fix lint issues

mkevins committed:
Fix lint issue

The class-level suppression did not seem to entirely satisfy CI, so this
adds suppressions at each use-site.

mkevins committed:
Add "unused" to lint warning suppression

Maybe the nth time is a charm?

mkevins committed:
Comment out code to satisfy CI -Werror

I might be missing something simple here, but several attempts at
suppression have not worked. I'm commented the code. These tests will be
refactored in a later PR.

mkevins committed:
More lint fun

mkevins committed:
Clean up lint suppresions

Finally found out that the correct suppresion in this case is:
"UNUSED_VARIABLE".

RenanLukas reviewed (approved):
Thanks for the changes, @mkevins . It works as expected :shipit:

mkevins commented:
Thank you for reviewing and testing! :tada:
